### PR TITLE
Spawn new Thread for the export [backend], auto-download when export done [frontend]

### LINKED ISSUE #1379: Export times out when projects are too large.
The project "Thai_FM_SP2021" fails to export and shows a 504 error in the browser console. From the server we saw that it successfully zips the 771 MB project, but then times out before `LiftController.cs` > `ExportLiftFile()` returns to the frontend. It turns out we were waiting on the process to complete, when we intended to have that run independently and use SignalR to signal the frontend when the export is complete and the download is ready.

### PR DESCRIPTION
**Pull request by imnasnainaec**
Fixes #1379 
Bypasses #937 by not requiring the user to press a button when the export is ready to download.

Also, uses random temp export folders to prevent concurrency issues when there are multiple simultaneous attempts to export the same project.

This change is [Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376)

### CODE REVIEW

**Review by jasonleenaylor — COMMENTED** [review]
This will all work. I have a strong inkling that we can write this with less code. Not sure if CreateLiftExportAndSignaly needs to return a Task, if not it might be possible to just call that and not await it to get the same result as the thread start...

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376#-:-MjfEVDG9T20StnCEYN7:b-2an4ez)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @johnthagen)

**Review by imnasnainaec — COMMENTED** [review]
I agree, it could work with less code. I added the Task return for the sake of a unit test that didn't have signal access to know when the export is done. But I'm open to a joint refactor of these functions and their unit tests.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376#-:-Mjfjax47HGrGAxffGHg:b-3sulhj)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @johnthagen)

**Comment by johnthagen** [comment]
___
*[Backend/Controllers/LiftController.cs, line 249 at r7](https://reviewable.io/reviews/sillsdev/thecombine/1376#-Mjns5FFDnUAta7KC9Wg:-Mjns5FFDnUAta7KC9Wh:bys4c3r) ([raw file](https://github.com/sillsdev/thecombine/blob/848d12a84441a7776a2115c7aa2b9adcd9132281/Backend/Controllers/LiftController.cs#L249)):*
> ```Smalltalk
>                 return BadRequest("No words to export.");
>             }
>             Thread exportThread = new Thread(async () => await CreateLiftExportThenSignal(projectId, userId));
> ```

Use `var exportThread`

**Comment by johnthagen** [comment]
___
*[Backend/Controllers/LiftController.cs, line 249 at r7](https://reviewable.io/reviews/sillsdev/thecombine/1376#-Mjnsy9k257zWZUtsaVA:-Mjnsy9k257zWZUtsaVB:bc6uri6) ([raw file](https://github.com/sillsdev/thecombine/blob/848d12a84441a7776a2115c7aa2b9adcd9132281/Backend/Controllers/LiftController.cs#L249)):*
> ```Smalltalk
>                 return BadRequest("No words to export.");
>             }
>             Thread exportThread = new Thread(async () => await CreateLiftExportThenSignal(projectId, userId));
> ```

Rider warns: "Avoid using 'async' lambda when delegate type returns 'void'.

Some research leads me here: https://devblogs.microsoft.com/pfxteam/potential-pitfalls-to-avoid-when-passing-around-async-lambdas/

In a nutshell, it seems like what can happen is that the Thread can suspend when it hits the first `await`. Looking at the code in this example the rest of the function isn't "doing anything" with the export but it might be better to fix this pattern anyway for future proofing.

Rider suggests:

```
            async void Start() => await CreateLiftExportThenSignal(projectId, userId);
            var exportThread = new Thread(Start);
```

**Comment by johnthagen** [comment]
___
*[Backend/Services/LiftService.cs, line 184 at r7](https://reviewable.io/reviews/sillsdev/thecombine/1376#-MjnuW5P9byZI6YNSFGa:-MjnuW5P9byZI6YNSFGb:bwzneom) ([raw file](https://github.com/sillsdev/thecombine/blob/848d12a84441a7776a2115c7aa2b9adcd9132281/Backend/Services/LiftService.cs#L184)):*
> ```Smalltalk
>             // Generate the zip dir.
>             var exportDir = FileStorage.GenerateLiftExportDirPath(projectId);
>             var liftExportDir = Path.Combine(exportDir, "LiftExport-" + Time.UtcNowFilesafe());
> ```

Rather than using the current time, what is normally done for temporary directories that must be unique is that a random string of characters is used. That way there is no possible race condition at all.

We could then remove the need for `UtcNowFilesafe()`.

**Comment by johnthagen** [comment]
___
*[Backend/Controllers/LiftController.cs, line 262 at r7](https://reviewable.io/reviews/sillsdev/thecombine/1376#-MjnxmftCYSkjA-Iyr_L:-MjnxmftCYSkjA-Iyr_M:bigskrs) ([raw file](https://github.com/sillsdev/thecombine/blob/848d12a84441a7776a2115c7aa2b9adcd9132281/Backend/Controllers/LiftController.cs#L262)):*
> ```Smalltalk
>             // Store the temporary path to the exported file for user to download later.
>             _liftService.StoreExport(userId, exportedFilepath);
>             await _notifyService.Clients.All.SendAsync("DownloadReady", userId);
> ```

It seems like the magic string `"DownloadReady"` should be defined somewhere as a `const`. Perhaps `CombineHub.cs`?

**Comment by johnthagen** [comment]
I agree the `async` inside the `Thread` isn't buying us anything because there isn't any other tasks that can cooperatively multi-task on that `Thread` since it only does one thing.

That being said, it doesn't seem like the `async` really complicates it too much either?

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376)* status: 1 of 14 files reviewed, 1 unresolved discussion (waiting on @jasonleenaylor and @johnthagen)

___
*[Backend/Controllers/LiftController.cs, line 249 at r7](https://reviewable.io/reviews/sillsdev/thecombine/1376#-Mjns5FFDnUAta7KC9Wg:-Mjo2PU76ZhkSGBCDhn1:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/848d12a84441a7776a2115c7aa2b9adcd9132281/Backend/Controllers/LiftController.cs#L249)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

Use `var exportThread`
</blockquote></details>

Done.

**Comment by johnthagen** [comment]
With the new `Task.Run` implementation, `async` within the `Task` will improve efficiency now.

**Review by johnthagen — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, 4 of 7 files at r4, 2 of 4 files at r6, 3 of 3 files at r7, 3 of 3 files at r10, 2 of 3 files at r11, 4 of 4 files at r12, 2 of 2 files at r13, 1 of 1 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @imnasnainaec)

**Comment by johnthagen** [comment]
___
*[Backend/Controllers/LiftController.cs, line 257 at r14](https://reviewable.io/reviews/sillsdev/thecombine/1376#-MjoCO6d22o4tlbh-apI:-MjoCO6d22o4tlbh-apJ:bv8rlmp) ([raw file](https://github.com/sillsdev/thecombine/blob/cb518036e2d475bf659ad8a05210482f5f11e7ad/Backend/Controllers/LiftController.cs#L257)):*
> ```Smalltalk
>         }
> 
>         internal async Task<bool> CreateLiftExportThenSignal(string projectId, string userId)
> ```

This can be `private`.

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376)* status: all files reviewed, 1 unresolved discussion (waiting on @johnthagen)

___
*[Backend/Controllers/LiftController.cs, line 257 at r14](https://reviewable.io/reviews/sillsdev/thecombine/1376#-MjoCO6d22o4tlbh-apI:-MjoClHXBsXfnY7o3gqT:bcximja) ([raw file](https://github.com/sillsdev/thecombine/blob/cb518036e2d475bf659ad8a05210482f5f11e7ad/Backend/Controllers/LiftController.cs#L257)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

This can be `private`.
</blockquote></details>

No, it's used in a Test.

**Comment by johnthagen** [comment]
___
*[Backend/Controllers/LiftController.cs, line 257 at r14](https://reviewable.io/reviews/sillsdev/thecombine/1376#-MjoCO6d22o4tlbh-apI:-MjoDH9s7LZllrFFZBSJ:b-udu5ty) ([raw file](https://github.com/sillsdev/thecombine/blob/cb518036e2d475bf659ad8a05210482f5f11e7ad/Backend/Controllers/LiftController.cs#L257)):*
<details><summary><i>Previously, imnasnainaec (D. Ror.) wrote…</i></summary><blockquote>

No, it's used in a Test.
</blockquote></details>

Done. I referenced the wrong method

**Comment by johnthagen** [comment]
___
*[Backend/Controllers/LiftController.cs, line 213 at r14](https://reviewable.io/reviews/sillsdev/thecombine/1376#-MjoDTGK-ZeBXxTMPshh:-MjoDTGK-ZeBXxTMPshi:bv8rlmp) ([raw file](https://github.com/sillsdev/thecombine/blob/cb518036e2d475bf659ad8a05210482f5f11e7ad/Backend/Controllers/LiftController.cs#L213)):*
> ```Smalltalk
> 
>         // These internal methods are extracted for unit testing
>         internal async Task<IActionResult> ExportLiftFile(string projectId, string userId)
> ```

This can be `private`.

**Review by imnasnainaec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376)* status: 15 of 16 files reviewed, 1 unresolved discussion (waiting on @johnthagen)

___
*[Backend/Controllers/LiftController.cs, line 213 at r14](https://reviewable.io/reviews/sillsdev/thecombine/1376#-MjoDTGK-ZeBXxTMPshh:-MjoFLLE0el8DDRudBik:b-896fix) ([raw file](https://github.com/sillsdev/thecombine/blob/cb518036e2d475bf659ad8a05210482f5f11e7ad/Backend/Controllers/LiftController.cs#L213)):*
<details><summary><i>Previously, johnthagen wrote…</i></summary><blockquote>

This can be `private`.
</blockquote></details>

Done.

**Review by johnthagen — APPROVED** [review]
> Reviewed 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1376)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @imnasnainaec)
